tachyon/crypto/commitments/fri/two_adic_fri_prover.h

```diff
@@ -0,0 +1,128 @@
+#include "tachyon/crypto/commitments/fri/two_adic_fri_config.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_proof.h"
+#include "tachyon/math/finite_fields/extension_field_traits_forward.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename ExtField, typename ChallengeMMCS, typename Challenger>
+CommitPhaseResult<ExtField, ChallengeMMCS> CommitPhase(
+    TwoAdicFriConfig<ChallengeMMCS>& config,
+    const std::vector<std::vector<ExtField>>& inputs, Challenger& challenger) {
+  using Commitment = typename ChallengeMMCS::Commitment;
+  using ProverData = typename ChallengeMMCS::ProverData;
+
+  std::vector<ExtField> folded = inputs[0];
+  std::vector<Commitment> commits;
+  std::vector<ProverData> data;
+
+  while (folded.size() > config.Blowup()) {
+    math::RowMajorMatrix<ExtField> leaves((folded.size() + 1) / 2, 2);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < folded.size(); ++i) {
+      leaves(i / 2, i % 2) = folded[i];
+    }
+
+    Commitment commit;
+    ProverData prover_data;
+    CHECK(config.mmcs.Commit({leaves}, &commit, &prover_data));
+    commits.push_back(commit);
+    data.push_back(prover_data);
+
+    challenger.ObserveContainer(commit);
+    const ExtField beta = challenger.template SampleExtElement<ExtField>();
+
+    leaves = config.mmcs.GetMatrices(prover_data).back();
+    folded = FoldMatrix(beta, leaves);
+    for (size_t i = 0; i < inputs.size(); ++i) {
+      if (inputs[i].size() == folded.size()) {
+        for (size_t j = 0; j < inputs[i].size(); ++j) {
+          folded[j] += inputs[i][j];
+        }
+      }
+    }
+  }
+  CHECK_EQ(folded.size(), config.Blowup());
+  ExtField& final_poly = folded[0];
+
+  OPENMP_PARALLEL_FOR(size_t i = 0; i < folded.size(); ++i) {
+    CHECK_EQ(folded[i], final_poly);
+  }
+  challenger.ObserveContainer(final_poly.ToBaseFields());
+
+  return {commits, data, final_poly};
+}
+
+template <typename F, typename MMCS>
+std::vector<CommitPhaseProofStep<F, MMCS>> AnswerQuery(
+    size_t index, TwoAdicFriConfig<MMCS>& config,
+    const std::vector<typename MMCS::ProverData>& commit_phase_commits) {
+  using Proof = typename MMCS::Proof;
+  std::vector<CommitPhaseProofStep<F, MMCS>> ret(commit_phase_commits.size());
+  for (size_t i = 0; i < commit_phase_commits.size(); ++i) {
+    size_t index_i = index >> i;
+    size_t index_i_sibling = index_i ^ 1;
+    size_t index_pair = index_i >> 1;
+    std::vector<std::vector<F>> opened_rows;
+    Proof opening_proof;
+    CHECK(config.mmcs.CreateOpeningProof(index_pair, commit_phase_commits[i],
+                                         &opened_rows, &opening_proof));
+    CHECK_EQ(opened_rows.size(), size_t{1});
+    CHECK_EQ(opened_rows[0].size(), size_t{2});
+    const F& sibling_value = opened_rows[0][index_i_sibling % 2];
+    ret[i] = CommitPhaseProofStep<F, MMCS>{sibling_value, opening_proof};
+  }
+  return ret;
```

Comment:
Please use `base::Map()`

Reply:
Used `CreateVector` as this is not mapping from a different vector.